client/network/src/persist_peers.rs

```diff
@@ -0,0 +1,308 @@
+use std::{
+	collections::{HashMap, HashSet},
+	fmt,
+	future::Future,
+	io,
+	path::{Path, PathBuf},
+	pin::Pin,
+	sync::Arc,
+	task::{Context, Poll},
+	time::{Duration, Instant},
+};
+
+use futures::FutureExt;
+use lru::LruCache;
+
+use sc_peerset::PeersetHandle;
+
+use crate::{Multiaddr, PeerId};
+
+type BoxedFuture<T> = Pin<Box<dyn Future<Output = T> + Send + 'static>>;
+type Never = std::convert::Infallible;
+type ProtocolType = String;
+
+const FLUSH_INTERVAL: Duration = Duration::from_secs(5);
+const PEER_ADDRS_CACHE_SIZE: usize = 100;
+
+pub struct PersistPeerAddrs {
+	paths: Arc<Paths>,
+	flushed_at: Instant,
+	protocols: HashMap<ProtocolType, LruCache<PeerId, HashSet<Multiaddr>>>,
+	busy: Option<BoxedFuture<Result<(), io::Error>>>,
+}
+
+impl PersistPeerAddrs {
+	pub fn load(dir: impl AsRef<Path>) -> Self {
+		let paths = Paths::new(dir, "peer-addrs");
+
+		let protocols = match persist_peer_addrs::load(&paths.path) {
+			Ok(restored) => restored,
+			Err(reason) => {
+				log::warn!("Failed to load peer addresses: {:?}", reason);
+				Default::default()
+			},
+		};
+
+		let protocols = protocols
+			.into_iter()
+			.map(|(protocol, entries)| {
+				let cache = entries.into_iter().rev().fold(
+					LruCache::new(PEER_ADDRS_CACHE_SIZE),
+					|mut acc, persist_peer_addrs::PeerEntry { peer_id, addrs }| {
+						if let Ok(peer_id) = peer_id.parse() {
+							acc.push(peer_id, addrs.into_iter().collect::<HashSet<_>>());
+						}
+						acc
+					},
+				);
+				(protocol, cache)
+			})
+			.collect();
+
+		Self { paths: Arc::new(paths), flushed_at: Instant::now(), protocols, busy: None }
+	}
+
+	pub fn report_peer_addr(
+		&mut self,
+		peer_id: &PeerId,
+		protocol: impl AsRef<[u8]>,
+		addr: &Multiaddr,
+	) {
+		let protocol = String::from_utf8(protocol.as_ref().to_owned()).expect(
+			"According to the `crate::discovery::protocol_name_from_protocol_id` \
+					and `<ProtocolId as AsRef<str>>` it's a correct UTF-8 string",
+		);
+
+		let entries = self
+			.protocols
+			.entry(protocol)
+			.or_insert_with(|| LruCache::new(PEER_ADDRS_CACHE_SIZE));
+		if let Some(peer_addrs) = entries.get_mut(peer_id) {
+			peer_addrs.insert(addr.to_owned());
+		} else {
+			entries.push(peer_id.to_owned(), [addr.to_owned()].into_iter().collect());
+		}
+	}
+
+	pub fn peer_addrs<'a>(
+		&'a mut self,
+		peer_id: &'a PeerId,
+		protocols: impl IntoIterator<Item = impl AsRef<[u8]> + 'a>,
+	) -> impl Iterator<Item = &'a Multiaddr> {
+		let protocols = protocols.into_iter().collect::<Vec<_>>();
+
+		self.protocols
+			.iter_mut()
```

Comment:
Can't this use `iter()`? (Or am I just blind and it does need `iter_mut()`?)

Reply:
```rust
.flat_map(|entries| entries.get(peer_id).into_iter())
```

this uses the LRU's `pub fn get<'a, Q>(&'a mut self, k: &Q) -> Option<&'a V>` method.